tikv/raftstore/snap_manager.go

```diff
@@ -0,0 +1,355 @@
+	}
+}
+
+type SnapManager struct {
+	base         string
+	snapSize     *int64
+	registryLock sync.RWMutex
+	registry     map[SnapKey][]SnapEntry
+	ch           chan<- *Msg
+	limiter      *IOLimiter
+	MaxTotalSize uint64
+}
+
+func NewSnapManager(path string, ch chan<- *Msg) *SnapManager {
+	return new(SnapManagerBuilder).Build(path, ch)
+}
+
+func (sm *SnapManager) init() error {
+	fi, err := os.Stat(sm.base)
+	if os.IsNotExist(err) {
+		err = os.MkdirAll(sm.base, 0600)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		return nil
+	} else if err != nil {
+		return errors.WithStack(err)
+	}
+	if !fi.IsDir() {
+		return errors.Errorf("%s should be a directory", sm.base)
+	}
+	fis, err := ioutil.ReadDir(sm.base)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	for _, fi := range fis {
+		if !fi.IsDir() {
+			name := fi.Name()
+			if strings.HasSuffix(name, tmpFileSuffix) {
+				err = os.Remove(filepath.Join(sm.base, name))
+				if err != nil {
+					return errors.WithStack(err)
+				}
+			} else if strings.HasSuffix(name, sstFileSuffix) {
+				atomic.AddInt64(sm.snapSize, fi.Size())
+			}
+		}
+	}
+	return nil
+}
+
+func (sm *SnapManager) ListIdleSnap() ([]SnapKeyWithSending, error) {
+	fis, err := ioutil.ReadDir(sm.base)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	results := make([]SnapKeyWithSending, 0, len(fis))
+	for _, fi := range fis {
+		if fi.IsDir() {
+			continue
+		}
+		name := fi.Name()
+		if !strings.HasSuffix(name, metaFileSuffix) {
+			continue
+		}
+		name = name[:len(name)-len(metaFileSuffix)]
+		var key SnapKeyWithSending
+		if strings.HasPrefix(name, snapGenPrefix) {
+			key.IsSending = true
+		}
+		numberStrs := strings.Split(name, "_")
+		if len(numberStrs) != 4 {
+			return nil, errors.Errorf("failed to parse file %s", name)
+		}
+		key.SnapKey.RegionID, err = strconv.ParseUint(numberStrs[1], 10, 64)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		key.SnapKey.Term, err = strconv.ParseUint(numberStrs[2], 10, 64)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		key.SnapKey.Index, err = strconv.ParseUint(numberStrs[3], 10, 64)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		sm.registryLock.RLock()
+		_, ok := sm.registry[key.SnapKey]
+		sm.registryLock.RUnlock()
+		if ok {
+			// Skip those registered snapshot.
+			continue
+		}
+		results = append(results, key)
+	}
+	sort.Slice(results, func(i, j int) bool {
+		keyI := &results[i].SnapKey
+		keyJ := &results[j].SnapKey
+		if keyI.RegionID == keyJ.RegionID {
+			if keyI.Term == keyJ.Term {
+				if keyI.Index == keyJ.Index {
+					return !results[i].IsSending
+				}
+				return keyI.Index < keyJ.Index
+			}
+			return keyI.Term < keyJ.Term
+		}
+		return keyI.RegionID < keyJ.RegionID
+	})
+	return results, nil
+}
+
+func (sm *SnapManager) HasRegistered(key SnapKey) bool {
+	sm.registryLock.RLock()
+	_, ok := sm.registry[key]
+	sm.registryLock.RUnlock()
```

Comment:
Why not use defer?

Reply:
I think It's ok to not defer if there is just one line of code.